src/iamserver/publicmessagehandler.cpp

```diff
@@ -0,0 +1,439 @@
+        (*permissions.mutable_permissions())[key.CStr()] = val.CStr();
+    }
+
+    *response->mutable_instance()    = instanceIdent;
+    *response->mutable_permissions() = permissions;
+
+    return grpc::Status::OK;
+}
+
+/***********************************************************************************************************************
+ * IAMPublicNodesService implementation
+ **********************************************************************************************************************/
+
+grpc::Status PublicMessageHandler::GetAllNodeIDs([[maybe_unused]] grpc::ServerContext* context,
+    [[maybe_unused]] const google::protobuf::Empty* request, iamproto::NodesID* response)
+{
+    LOG_DBG() << "Public message handler. Process get all node IDs";
+
+    aos::StaticArray<aos::StaticString<aos::cNodeIDLen>, aos::cMaxNumNodes> nodeIDs;
+
+    if (auto err = mNodeManager->GetAllNodeIds(nodeIDs); !err.IsNone()) {
+        LOG_ERR() << "Failed to get all node IDs: " << AOS_ERROR_WRAP(err);
+
+        return grpc::Status(grpc::StatusCode::INTERNAL, "Failed to get all node IDs");
+    }
+
+    if (nodeIDs.Size() == 0) {
+        return grpc::Status(grpc::StatusCode::INTERNAL, "Empty node IDs");
+    }
+
+    // Sort node IDs to have main node ID last
+    nodeIDs.Sort([mainNodeId = mNodeInfo.mID](const auto& a, const auto&) { return a == mainNodeId; });
+
+    for (const auto& id : nodeIDs) {
+        response->add_ids(id.CStr());
+    }
+
+    return grpc::Status::OK;
+}
+
+grpc::Status PublicMessageHandler::GetNodeInfo([[maybe_unused]] grpc::ServerContext* context,
+    [[maybe_unused]] const iamproto::GetNodeInfoRequest* request, iamproto::NodeInfo* response)
+{
+    LOG_DBG() << "Public message handler. Process get node info: id = " << request->node_id().c_str();
+
+    aos::NodeInfo nodeInfo;
+
+    if (auto err = mNodeManager->GetNodeInfo(request->node_id().c_str(), nodeInfo); !err.IsNone()) {
+        LOG_ERR() << "Failed to get node info: " << AOS_ERROR_WRAP(err);
+
+        return grpc::Status(grpc::StatusCode::INTERNAL, "Failed to get node info");
+    }
+
+    ConvertToProto(nodeInfo, *response);
+
+    return grpc::Status::OK;
+}
+
+grpc::Status PublicMessageHandler::SubscribeNodeChanged([[maybe_unused]] grpc::ServerContext* context,
+    [[maybe_unused]] const google::protobuf::Empty* request, grpc::ServerWriter<iamproto::NodeInfo>* writer)
+{
+    LOG_DBG() << "Public message handler. Got subscribe node changed request";
+
+    return mNodeChangedController.HandleStream(context, writer);
+}
+
+grpc::Status PublicMessageHandler::RegisterNode(grpc::ServerContext*                            context,
+    grpc::ServerReaderWriter<::iamproto::IAMIncomingMessages, ::iamproto::IAMOutgoingMessages>* stream)
+{
+    LOG_DBG() << "Public message handler. Process register node";
+
+    using aos::NodeStatus;
+    using aos::NodeStatusEnum;
+
+    static const std::vector<NodeStatus> allowedStatuses = {NodeStatusEnum::eUnprovisioned};
```

Comment:
for my opinion will be better to move in const section, underhood lazy initialization used mutex